asf_heat_pump_suitability/analysis/exploratory/create_full_dataset/assign_cluster_suitability_and_feasibility.py

```diff
@@ -0,0 +1,313 @@
+
+
+def create_df_feasibility_scoring(
+    df: pl.DataFrame,
+    features: list = config.features,
+    expected_tech_types: set = config.expected_tech_types,
+    weights: dict = config.weights,
+) -> pl.DataFrame:
+    """
+    Computes the feasibility score for each tech type and cluster in the dataset.
+
+    Tech types can include:
+    - Individual Air Source Heat Pump (ASHP)
+    - Collective purchasing of ASHP
+    - Shared Ground Loop (SGL)
+    - Heat Network (HN)
+
+    Args:
+        df (pl.DataFrame): DataFrame containing the dataset with feature values for each UPRN.
+        features (list, optional): List of features used in the feasibility scoring.
+            Defaults to a predefined list.
+        expected_tech_types (set, optional): Set of expected technology types for feasibility scoring.
+        weights (dict, optional): Dictionary containing the weights for each feature used to compute feasibility for specific tech types.
+            in the format {'tech_type_1': {'feature_1': weight_1, 'feature_2': weight_2, ...}, ..., 'tech_type_n': {...}}
+
+    Raises:
+        ValueError: if tech types or features provided in weights do not exist.
+
+    Returns:
+        pl.DataFrame: DataFrame with aggregated feasibility scores for each cluster.
+    """
+    if set(weights.keys()) != expected_tech_types:
+        incorrect_techs = [w for w in weights.keys() if w not in expected_tech_types]
+        missing_techs = [t for t in expected_tech_types if t not in weights.keys()]
+        raise ValueError(
+            f"There are incorrect or missing keys in the 'weights' dictionary.\n Incorrect techs: {incorrect_techs}\n Missing techs: {missing_techs}"
+        )
+
+    for tech in weights.keys():
+        tech_features = set(weights.get(tech).keys())
+        if not tech_features.issubset(set(features + ["cluster_size"])):
+            do_not_exist = [
+                f for f in tech_features if f not in set(features + ["cluster_size"])
+            ]
+            raise ValueError(
+                f"{tech}: The features you're providing weights for do not exist:\n{do_not_exist}"
+            )
+
+    # Create feasibility expressions for all tech types and store as list
+    tech_feasibility_scores = [
+        calculate_feasibility_expression(tech_specific_weights=weights.get(tech)).alias(
+            tech + "_feasibility"
+        )
+        for tech in expected_tech_types
+    ]
+
+    # Add feasibility scores as new columns to cluster_stats
+    df = df.with_columns(tech_feasibility_scores)
+
+    return df
+
+
+if __name__ == "__main__":
+    plymouth_data = pl.read_parquet(
+        "s3://asf-heat-pump-suitability/exploration/spatial_clustering_plymouth/plymouth_uprn_with_features.parquet"
+    )
+
+    feasibility_scoring_data = prepare_df_for_feasibility_scoring(
+        df=plymouth_data,
+        features=config.features,
+        anchor_loads_threshold=500,
+        outdoor_space_threshold=0,
```

Comment:
we should change these defaults to values that make sense before running the script